app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt

```diff
@@ -0,0 +1,201 @@
+
+    private lateinit var keyStore: KeyStore
+    private lateinit var keyGenerator: KeyGenerator
+    private var cancellationSignal: CancellationSignal? = null
+    private var selfCancelled = false
+
+    val authState: ReplaySubject<AuthenticationState> = ReplaySubject.create()
+
+    companion object {
+        val shared = FingerprintStore()
+        private const val ANDROID_KEY_STORE = "AndroidKeyStore"
+        private const val DEFAULT_KEY = "DEFAULT_KEY"
+    }
+
+    sealed class AuthenticationState {
+        object Succeeded : AuthenticationState()
+        object Failed : AuthenticationState()
+        data class Error(val error: String?) : AuthenticationState()
+    }
+
+    init {
+        dispatcher.register.filterByType(AuthenticationAction::class.java)
+            .subscribe {
+                when (it) {
+                    is AuthenticationAction.StartListening -> initFingerprint()
+                    is AuthenticationAction.StopListening -> stopListening()
+                }
+            }
+            .addTo(compositeDisposable)
+    }
+
+    fun apply(manager: FingerprintManager) {
+        fingerprintManager = manager
+    }
+
+    fun isFingerprintAuthAvailable(): Boolean {
+        return fingerprintManager.isHardwareDetected && fingerprintManager.hasEnrolledFingerprints()
+    }
+
+    private fun initFingerprint() {
+        setupKeyStoreAndKeyGenerator()
+        createKey(DEFAULT_KEY)
+        val (defaultCipher: javax.crypto.Cipher, cipherNotInvalidated: javax.crypto.Cipher) = setupCiphers()
+        initCipher(defaultCipher, DEFAULT_KEY)
+
+        startListening(FingerprintManager.CryptoObject(defaultCipher))
+    }
+
+    private fun startListening(cryptoObject: FingerprintManager.CryptoObject) {
+        if (!isFingerprintAuthAvailable()) return
+        cancellationSignal = CancellationSignal()
+        selfCancelled = false
+        fingerprintManager.authenticate(cryptoObject, cancellationSignal, 0, AuthenticationCallback(), null)
+    }
+
+    private fun stopListening() {
+        cancellationSignal?.also {
+            selfCancelled = true
+            it.cancel()
+        }
+        cancellationSignal = null
+    }
+
+    private fun setupKeyStoreAndKeyGenerator() {
+        try {
+            keyStore = KeyStore.getInstance(ANDROID_KEY_STORE)
+        } catch (e: KeyStoreException) {
+            throw RuntimeException("Failed to get an instance of KeyStore", e)
+        }
+
+        try {
+            keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEY_STORE)
+        } catch (e: Exception) {
+            when (e) {
+                is NoSuchAlgorithmException,
+                is NoSuchProviderException ->
+                    throw RuntimeException("Failed to get an instance of KeyGenerator", e)
+                else -> throw e
+            }
+        }
+    }
+
+    private fun setupCiphers(): Pair<Cipher, Cipher> {
```

Comment:
All of this key generation work can use the `DataProtect.Keystore` class instead of us doing the heavy lifting ourselves

Reply:
I'm going to make a follow-up story to address this because the right way to hook these up (observables? simple method calls?) is not obvious to me, and I'd rather get them both landed sooner than later.

Reply:
:+1: